### PR TITLE
Reuse incremental JIT compilation for --image-codegen

### PR DESCRIPTION
**Pull request by pchintalapudi**
We don't need to merge all of the workqueue modules when performing compilation with `--image-codegen` set, we just need the global variable initializers to be defined before they're used in one of the modules. Therefore we can do this by compiling all of the global variable initializers upfront, so that later references will link them properly.

### CODE REVIEW

File: src/jitlayers.cpp [hunk 1]
```diff
@@ -211,46 +229,53 @@ static jl_callptr_t _jl_compile_codeinst(
...
             jl_ExecutionEngine->addModule(orc::ThreadSafeModule(std::move(params._shared_module), params.tsctx));
-        if (!params.imaging) {
-            StringMap<orc::ThreadSafeModule*> NewExports;
+
+        // In imaging mode, we can't inline global variable initializers in order to preserve
+        // the fiction that we don't know what loads from the global will return. Thus, we
+        // need to emit a separate module for the globals before any functions are compiled,
+        // to ensure that the globals are defined when they are compiled.
+        if (params.imaging) {
+            jl_ExecutionEngine->addModule(jl_get_globals_module(params.tsctx, params.imaging, params.DL, params.TargetTriple, params.globals));
```
Comment (vtjnash) [inline src/jitlayers.cpp:238]:
```suggestion
            auto lock = params.tsctx.getLock();
            for (auto &global : params.globals) {
                auto GV = global.second;
                GV->setName(GV->getName() + "." + unique_counter++);
                addGlobalMapping(GV->getName(), global.first);
            }
```

**Review by vtjnash — COMMENTED** [review]
LGTM, other than `jl_get_globals_module` which needs to mangle names to something unique in the current `jl_ExecutionEngine`.

**Comment by pchintalapudi** [comment]
That's actually handled elsewhere in `julia_pgv`, which uses the `globalUniqueGeneratedNames` counter when in imaging mode. Should we move that uniquing logic here?

**Comment by vtjnash** [comment]
Ah right, so you can remove that extra setName call then for now, but still simplify that loop substantially by using `addGlobalMapping` directly. Eventually we should remove globalUniqueGeneratedNames since it can be instead handled better by the JIT like this, as a strictly local property.
```c++
            for (auto &global : params.globals)
                addGlobalMapping(global.second->getName(), global.first);
```

The `addGlobalMapping` also isn't quite the same as imaging mode yet anyways, since we should actually append them into an array in the module itself later with internal linkage, so that they are all completely anonymous, then do a simple copy afterwards to fill in that array when doing the final linking step to populate those values (like we do in staticdata.c).

That should be some of the last steps needed to make codegen itself pure, and make native code caching feasible based on the IR or bitcode hash.

**Comment by pchintalapudi** [comment]
I don't think addGlobalMapping is helpful here, we want the pointer to be the value stored in the global rather than the address of the global. We still need to allocate memory to load from in the IR.

**Comment by vtjnash** [comment]
Ah, right. That probably gets us even closer to the version where we actually do the same representation as image_codegen though, since we would already make the equivalent array for it:
```c++
            void **globals = (void**)malloc(sizeof(void*) * params.globals.size());
            size_t i = 0;
            for (auto &global : params.globals) {
                globals[i] = global.first;
                addGlobalMapping(global.second->getName(), &globals[i]);
                i++;
            }
```
Then all the remains is converting all of the `global.second` objects into offsets into this array, like is done in aotcompile to build the fvars list.

**Comment by pchintalapudi** [comment]
I think the better route to go about this is to convert imaging mode to use direct pointers instead of offsets, and make the linker patch in those references at load time.

**Pull request merged**